src/anomalib/metrics/per_image/pimo_numpy.py

```diff
@@ -0,0 +1,410 @@
+"""Per-Image Overlap curve (PIMO, pronounced pee-mo) and its area under the curve (AUPIMO).
+
+Details: `anomalib.metrics.per_image.pimo`.
+"""
+
+# Original Code
+# https://github.com/jpcbertoldo/aupimo
+#
+# Modified
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from enum import Enum
+
+import numpy as np
+from numpy import ndarray
+
+from . import _validate, binclf_curve_numpy
+from .binclf_curve_numpy import BinclfAlgorithm, BinclfThreshsChoice
+
+logger = logging.getLogger(__name__)
+
+# =========================================== CONSTANTS ===========================================
+
+
+class PIMOSharedFPRMetric(Enum):
+    """Shared FPR metric (x-axis of the PIMO curve)."""
+
+    MEAN_PERIMAGE_FPR: str = "mean-per-image-fpr"
+
+
+# =========================================== AUX ===========================================
+
+
+def _images_classes_from_masks(masks: ndarray) -> ndarray:
+    """Deduce the image classes from the masks."""
+    _validate.is_masks(masks)
+    return (masks == 1).any(axis=(1, 2)).astype(np.int32)
+
+
+# =========================================== ARGS VALIDATION ===========================================
+
+
+def _validate_has_at_least_one_anomalous_image(masks: ndarray) -> None:
+    image_classes = _images_classes_from_masks(masks)
+    if (image_classes == 1).sum() == 0:
+        msg = "Expected at least one ANOMALOUS image, but found none."
+        raise ValueError(msg)
+
+
+def _validate_has_at_least_one_normal_image(masks: ndarray) -> None:
+    image_classes = _images_classes_from_masks(masks)
+    if (image_classes == 0).sum() == 0:
+        msg = "Expected at least one NORMAL image, but found none."
+        raise ValueError(msg)
+
+
+def _joint_validate_threshs_shared_fpr(threshs: ndarray, shared_fpr: ndarray) -> None:
+    if threshs.shape[0] != shared_fpr.shape[0]:
+        msg = (
+            "Expected `threshs` and `shared_fpr` to have the same number of elements, "
+            f"but got {threshs.shape[0]} != {shared_fpr.shape[0]}"
+        )
+        raise ValueError(msg)
+
+
+# =========================================== PIMO ===========================================
```

Comment:
Looks like these validators are used only once. Is it needed to create a function for this logic? Maybe just validate it within the place, which is `thresh_at_shared_fpr_level` in this case?